flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureRateRestartBackoffTimeStrategy.java

```diff
@@ -0,0 +1,145 @@
+import static org.apache.flink.configuration.ConfigConstants.RESTART_BACKOFF_TIME_STRATEGY_FAILURE_RATE_MAX_FAILURES_PER_INTERVAL;
+import static org.apache.flink.configuration.ConfigConstants.RESTART_BACKOFF_TIME_STRATEGY_RESTART_BACKOFF_TIME;
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Restart strategy which can restart when failure rate is not exceeded.
+ */
+public class FailureRateRestartBackoffTimeStrategy implements RestartBackoffTimeStrategy {
+
+	private final long failuresIntervalMS;
+
+	private final long backoffTimeMS;
+
+	private final int maxFailuresPerInterval;
+
+	private final ArrayDeque<Long> failureTimestamps;
+
+	private final String strategyString;
+
+	public FailureRateRestartBackoffTimeStrategy(int maxFailuresPerInterval, long failuresInterval, long backoffTimeMS) {
+
+		checkArgument(maxFailuresPerInterval > 0, "Maximum number of restart attempts per time unit must be greater than 0.");
+		checkArgument(failuresInterval > 0, "Failures interval must be greater than 0 ms.");
+		checkArgument(backoffTimeMS >= 0, "Backoff time must be at least 0 ms.");
+
+		this.failuresIntervalMS = failuresInterval;
+		this.backoffTimeMS = backoffTimeMS;
+		this.maxFailuresPerInterval = maxFailuresPerInterval;
+		this.failureTimestamps = new ArrayDeque<>(maxFailuresPerInterval);
+		this.strategyString = generateStrategyString();
+	}
+
+	@Override
+	public boolean canRestart() {
+		if (isFailureTimestampsQueueFull()) {
+			Long now = System.currentTimeMillis();
+			Long earliestFailure = failureTimestamps.peek();
+
+			return (now - earliestFailure) > failuresIntervalMS;
+		} else {
+			return true;
+		}
+	}
+
+	@Override
+	public long getBackoffTime() {
+		return backoffTimeMS;
+	}
+
+	@Override
+	public void notifyFailure(Throwable cause) {
+		if (isFailureTimestampsQueueFull()) {
+			failureTimestamps.remove();
+		}
+		failureTimestamps.add(System.currentTimeMillis());
+	}
+
+	@Override
+	public String toString() {
+		return strategyString;
+	}
+
+	private boolean isFailureTimestampsQueueFull() {
+		return failureTimestamps.size() >= maxFailuresPerInterval;
+	}
+
+	private String generateStrategyString() {
+		StringBuilder buf = new StringBuilder("FailureRateRestartBackoffTimeStrategy(");
+		buf.append("FailureRateRestartBackoffTimeStrategy(failuresIntervalMS=");
+		buf.append(failuresIntervalMS);
+		buf.append(",backoffTimeMS=");
+		buf.append(backoffTimeMS);
+		buf.append(",maxFailuresPerInterval=");
+		buf.append(maxFailuresPerInterval);
+		buf.append(")");
+
+		return buf.toString();
+	}
+
+	public static FailureRateRestartBackoffTimeStrategyFactory createFactory(final Configuration configuration) {
+		return new FailureRateRestartBackoffTimeStrategyFactory(configuration);
+	}
+
+	/**
+	 * The factory for creating {@link FailureRateRestartBackoffTimeStrategy}.
+	 */
+	public static class FailureRateRestartBackoffTimeStrategyFactory implements RestartBackoffTimeStrategy.Factory {
```

Comment:
seems this class will only be used in `FailureRateRestartBackoffTimeStrategy`, the Java access modifiers may be changed?

Reply:
It will be used in `RestartBackoffTimeStrategyFactoryLoader`of another PR.